### ISSUE TITLE
support for Wan2.1-1.3B / 5B

### ISSUE THREAD

**Issue by johndpope** [WeChatCV/Stand-In#21]
i attempt to hotwire but doesnt have any impact - did you train checkpoint for 1.3 or 5b would also be better.
https://huggingface.co/Wan-AI/Wan2.2-TI2V-5B

```python
print("Loading model, please wait...")
try:
    ANTELOPEV2_PATH = "checkpoints/antelopev2"
    BASE_MODEL_PATH = "checkpoints/base_model_1_3b/"
    LORA_MODEL_PATH = "checkpoints/Stand-In/Stand-In_wan2.1_T2V_14B_ver1.0.ckpt"

    if not os.path.exists(ANTELOPEV2_PATH):
        raise FileNotFoundError(
            f"AntelopeV2 checkpoint not found at: {ANTELOPEV2_PATH}"
        )
    if not os.path.exists(BASE_MODEL_PATH):
        raise FileNotFoundError(f"Base model not found at: {BASE_MODEL_PATH}")
    if not os.path.exists(LORA_MODEL_PATH):
        raise FileNotFoundError(f"LoRA model not found at: {LORA_MODEL_PATH}")

    face_processor = FaceProcessor(antelopv2_path=ANTELOPEV2_PATH)
    pipe = load_wan_pipe(base_path=BASE_MODEL_PATH, torch_dtype=torch.bfloat16)
    set_stand_in_with_adapter(pipe, model_path=LORA_MODEL_PATH)
    print("Model loaded successfully!")
except Exception as e:
    print(f"Model loading failed: {e}")
    with gr.Blocks() as demo:
        gr.Markdown("# Error: Model Loading Failed")
        gr.Markdown(f"""
        Please check the following:
        1.  Make sure the checkpoint files are placed in the correct directory.
        2.  Ensure all dependencies are properly installed.
        3.  Check the console output for detailed error information.
        
        **Error details**: {e}
        """)
    demo.launch()
    exit()
```

```python

import torch
import torch.nn as nn
import torch.nn.functional as F


class DimensionAdapter(nn.Module):
    """Adapter to project between different model dimensions."""
    
    def __init__(self, source_dim, target_dim):
        super().__init__()
        self.source_dim = source_dim
        self.target_dim = target_dim
        
        # Create projection layers
        self.down_proj = nn.Linear(source_dim, target_dim, bias=False)
        self.up_proj = nn.Linear(target_dim, source_dim, bias=False)
        
        # Initialize with identity-like mapping where possible
        with torch.no_grad():
            # For down projection (source_dim -> target_dim)
            # Weight shape: [target_dim, source_dim]
            min_dim = min(source_dim, target_dim)
            self.down_proj.weight.data.zero_()
            self.down_proj.weight.data[:min_dim, :min_dim] = torch.eye(min_dim)
            
            # For up projection (target_dim -> source_dim)
            # Weight shape: [source_dim, target_dim]
            self.up_proj.weight.data.zero_()
            self.up_proj.weight.data[:min_dim, :min_dim] = torch.eye(min_dim)
    
    def adapt_down_weight(self, weight):
        """Adapt a down projection weight from source to target dimension."""
        # weight shape: [lora_rank, source_dim] -> [lora_rank, target_dim]
        lora_rank = weight.shape[0]
        if weight.shape[1] == self.source_dim:
            # Simply truncate or pad the weight matrix
            adapted = torch.zeros(lora_rank, self.target_dim, dtype=weight.dtype, device=weight.device)
            min_dim = min(self.source_dim, self.target_dim)
            adapted[:, :min_dim] = weight[:, :min_dim]
            return adapted
        return weight
    
    def adapt_up_weight(self, weight):
        """Adapt an up projection weight from source to target dimension."""
        # weight shape: [source_dim, lora_rank] -> [target_dim, lora_rank]
        lora_rank = weight.shape[1]
        if weight.shape[0] == self.source_dim:
            # Simply truncate or pad the weight matrix
            adapted = torch.zeros(self.target_dim, lora_rank, dtype=weight.dtype, device=weight.device)
            min_dim = min(self.source_dim, self.target_dim)
            adapted[:min_dim, :] = weight[:min_dim, :]
            return adapted
        return weight


def set_stand_in_with_adapter(pipe, train=False, model_path=None):
    """Set up Stand-In with dimension adaptation for compatibility."""
    # Initialize LoRA layers
    for block in pipe.dit.blocks:
        block.self_attn.init_lora(train)
    
    if model_path is not None:
        print(f"Loading Stand-In weights from: {model_path}")
        # Get model dimension from first block
        model_dim = pipe.dit.blocks[0].self_attn.q_loras.down.weight.shape[1]
        print(f"Model dimension: {model_dim}")
        
        # Load with dimension adaptation
        load_lora_weights_with_adapter(pipe, model_path, model_dim)


def load_lora_weights_with_adapter(pipe, ckpt_path, target_dim, strict=False):
    """Load LoRA weights with dimension adaptation."""
    ckpt = torch.load(ckpt_path, map_location="cpu")
    state_dict = ckpt.get("state_dict", ckpt)
    
    # Detect source dimension from checkpoint
    source_dim = None
    for k, v in state_dict.items():
        if "q_loras.down.weight" in k:
            source_dim = v.shape[1]
            break
    
    if source_dim is None:
        raise ValueError("Could not detect source dimension from checkpoint")
    
    print(f"Source dimension (from checkpoint): {source_dim}")
    print(f"Target dimension (current model): {target_dim}")
    
    # Create adapter if dimensions mismatch
    adapter = None
    if source_dim != target_dim:
        print(f"Creating dimension adapter: {source_dim} -> {target_dim}")
        adapter = DimensionAdapter(source_dim, target_dim)
        adapter.eval()  # Set to eval mode
    
    model = {}
    for i, block in enumerate(pipe.dit.blocks):
        prefix = f"blocks.{i}.self_attn."
        attn = block.self_attn
        for name in ["q_loras", "k_loras", "v_loras"]:
            for sub in ["down", "up"]:
                key = f"{prefix}{name}.{sub}.weight"
                if hasattr(getattr(attn, name), sub):
                    model[key] = getattr(getattr(attn, name), sub).weight
                else:
                    if strict:
                        raise KeyError(f"Missing module: {key}")
    
    # Load weights with adaptation
    loaded_keys = []
    skipped_keys = []
    
    for k, param in state_dict.items():
        if k in model:
            target_param = model[k]
            
            if adapter is not None:
                # Apply dimension adaptation
                if "down.weight" in k:
                    # Adapt down projection: [lora_rank, source_dim] -> [lora_rank, target_dim]
                    if param.shape[1] == source_dim and target_param.shape[1] == target_dim:
                        adapted_weight = adapter.adapt_down_weight(param)
                        if adapted_weight.shape == target_param.shape:
                            target_param.data.copy_(adapted_weight)
                            loaded_keys.append(k)
                        else:
                            print(f"Shape mismatch after adaptation: {k}")
                            skipped_keys.append(k)
                    else:
                        # Try direct copy if dimensions match
                        if param.shape == target_param.shape:
                            target_param.data.copy_(param)
                            loaded_keys.append(k)
                        else:
                            print(f"Unexpected shape: {k} | {param.shape} vs {target_param.shape}")
                            skipped_keys.append(k)
                            
                elif "up.weight" in k:
                    # Adapt up projection: [source_dim, lora_rank] -> [target_dim, lora_rank]
                    if param.shape[0] == source_dim and target_param.shape[0] == target_dim:
                        adapted_weight = adapter.adapt_up_weight(param)
                        if adapted_weight.shape == target_param.shape:
                            target_param.data.copy_(adapted_weight)
                            loaded_keys.append(k)
                        else:
                            print(f"Shape mismatch after adaptation: {k}")
                            skipped_keys.append(k)
                    else:
                        # Try direct copy if dimensions match
                        if param.shape == target_param.shape:
                            target_param.data.copy_(param)
                            loaded_keys.append(k)
                        else:
                            print(f"Unexpected shape: {k} | {param.shape} vs {target_param.shape}")
                            skipped_keys.append(k)
            else:
                # No adaptation needed, direct copy
                if param.shape == target_param.shape:
                    target_param.data.copy_(param)
                    loaded_keys.append(k)
                else:
                    if strict:
                        raise ValueError(
                            f"Shape mismatch: {k} | {target_param.shape} vs {param.shape}"
                        )
                    else:
                        skipped_keys.append(k)
        else:
            if strict:
                raise KeyError(f"Unexpected key in ckpt: {k}")
    
    print(f"Successfully loaded {len(loaded_keys)} parameters")
    if skipped_keys:
        print(f"Skipped {len(skipped_keys)} incompatible parameters")
    
    return adapter  # Return adapter in case it needs to be saved

```

**Comment by KBRASK** [comment]
Thank you for your suggestion! I will support wan2.2 5B in the future. But given that most community users are currently using wan2.2 T2V A14B, I will first support the adaptation of this model and then proceed to train other models.

**Comment by johndpope** [comment]
i did notice you did release the loss function in code - so its possible to train. 
i heard good things with 5B - are people running 14b on consumer hardware? im on 5090 - but get OOM.
